File: Java/src/com/dbc/algorithm/Stack/NextGreaterElementII.java
package com.dbc.algorithm.Stack;

import java.util.Arrays;
import java.util.Stack;

public class NextGreaterElementII {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < 2 * n - 1; i++) {
            while (!stack.isEmpty() && nums[stack.lastElement()] < nums[i % n]) {
                result[stack.pop()] = nums[i % n];
            }
            stack.push(i % n);
        }
        return result;
    }
}

File: Python/stack/next_greater_element_ii.py
from typing import List

def nextGreaterElements(nums: List[int]) -> List[int]:
    n = len(nums)
    ret = [-1] * n
    stk = list()

    for i in range(n * 2 - 1):
        while stk and nums[stk[-1]] < nums[i % n]:
            ret[stk.pop()] = nums[i % n]
        stk.append(i % n)
    
    return ret
